### ISSUE TITLE
DataLoader: collate_fn leads to high memory RAM memory usage

### ISSUE THREAD

**Issue by pantaleo18** [pytorch/pytorch#176166]
### Setup
OS: Windows 11 Enterprise
Pytorch Version: 2.9.1+cu128
Platform: Python 3.10

In snapshots, the script was running in **debug mode** (VSCode, Python Debugger Extension)

### Description
In my custom dataset, the average datapoint size is about 500 MB. Hence, considering `prefetch_factor = 2`, the expected memory usage per worker is approximately 1 GB. 

#### `collate_fn` = None

```
    train_loader = DataLoader(
        train_dataset,
        batch_size=1,
        shuffle=True,
        num_workers=4,
        pin_memory=False,
        prefetch_factor=2,
        persistent_workers=True,
        drop_last=False,
        worker_init_fn=seed_worker,
        collate_fn=None
    )
```

<img width="2000" height="723" alt="Image" src="https://github.com/user-attachments/assets/9dba7158-7671-443f-a39c-8516f4990652" />

```
[WORKER 1] PID=27328 | num_workers=4
===================================================
PROCESS MEMORY REPORT
PID: 27328 | PPID: 5668
===================================================
OS MEMORY (Task Manager view)
RSS (used RAM)        : 1296.01 MB
VMS (reserved)        : 2451.24 MB
Shared memory         : 0.00 MB
--------------------------------------------------------------------------------
PYTHON OBJECTS TOTAL  : 435070
--------------------------------------------------------------------------------
TOP 5 OBJECT CATEGORIES BY MEMORY:
torch.Storage             | count:      12 | memory:     599.13 MB
dict                      | count:   60899 | memory:      28.57 MB
function                  | count:   91561 | memory:      11.72 MB
set                       | count:    7022 | memory:       3.67 MB
tuple                     | count:   63437 | memory:       3.10 MB
--------------------------------------------------------------------------------
Explained memory      : 659.70 MB
Unexplained memory    : 636.30 MB
==================================================
```

#### `collate_fn = mil_pad_collate`
RAM usage increases linearly up to 22 GB (~7 GB per worker) during the first full iteration over the dataset. When `persistent_workers=True`, RAM usage grows only during the first epoch and remains stable afterwards.. Hence, subsequent epochs do not allocate further memory. 

```
    train_loader = DataLoader(
        train_dataset,
        batch_size=1,
        shuffle=True,
        num_workers=4,
        pin_memory=False,
        prefetch_factor=2,
        persistent_workers=True,
        drop_last=False,
        worker_init_fn=seed_worker,
        collate_fn=mil_pad_collate
    )
```

<img width="2000" height="721" alt="Image" src="https://github.com/user-attachments/assets/9958d957-9d67-480f-ba2e-1a9b0f51daf5" />

```
==============================================
PROCESS MEMORY REPORT
PID: 27328 | PPID: 36108
================================================
OS MEMORY (Task Manager view)
RSS (used RAM)        : 6212.75 MB
VMS (reserved)        : 7474.27 MB
Shared memory         : 0.00 MB
--------------------------------------------------------------------------------
PYTHON OBJECTS TOTAL  : 435096
--------------------------------------------------------------------------------
TOP 3 OBJECT CATEGORIES BY MEMORY:
torch.Storage             | count:      42 | memory:    2478.64 MB
torch.Tensor              | count:       3 | memory:     178.02 MB
dict                      | count:   60901 | memory:      28.54 MB
--------------------------------------------------------------------------------
Explained memory      : 2717.76 MB
Unexplained memory    : 3494.99 MB
==============================================
```

### Help Request
Can someone explain why the DataLoader allocates several additional gigabytes of RAM per worker when using a custom collate_fn, even with batch_size=1?

### CODE
This section shows the code linked to this issue/bug.

#### Dataset
```
class FeatureBagsDataset(Dataset):
    """
    PyTorch Dataset for loading pre-extracted feature bags stored in HDF5 files,
    indexed by a manifest DataFrame.

    Each sample corresponds to a single slide WSI and returns:
    - a bag of feature vectors
    - its associated label
    - optional patch coordinates

    This dataset is designed for Multiple Instance Learning (MIL) workflows.

    File Organization
    -----------------
    Feature files must be stored as HDF5 files with the following naming convention:

        <slide_id>.h5

    where ``slide_id`` matches exactly the value stored in the manifest DataFrame.

    The dataset accepts a list of directories ``data_dirs`` where it will search for
    the HDF5 files. Only these directories are searched; no recursive search is
    performed inside subfolders.

    HDF5 File Structure
    -------------------
    Each HDF5 file is expected to contain the following datasets:

    - ``features`` (default key): array of shape (N, D)
        Bag of N feature vectors of dimensionality D.
    - ``coords`` (default key, optional): array of shape (N, 2)
        Spatial coordinates associated with each feature vector.

    If the ``coords`` dataset is missing, ``coords`` will be returned as ``None``.

    Manifest DataFrame
    ------------------
    The input ``manifest_df`` must contain at least the following columns:

    - ``slide_id`` (or the column specified by ``slide_col``):
        Unique identifier for each slide. This value must match the HDF5 filename
        (without extension).

    - ``label`` (or the column specified by ``label_col``):
        Encoded class label as an integer (e.g., 0, 1, 2, 3, ...).
        These labels are intended for direct use in model training.

    Optionally, the manifest may also contain a column with human-readable class names:

    - ``class`` (or the column specified by ``class_col``):
        String representation of the true semantic class (e.g., "tumor", "normal",
        "grade_1"). This column is not used for training but can be useful for analysis,
        visualization, or debugging.

    If ``class_col`` is not provided or if a value is missing (NaN), the dataset will
    return ``None`` for the class name.

    Loading Behavior
    ----------------
    Feature data are loaded lazily on demand inside ``__getitem__``. No HDF5 files
    are kept open between calls. This makes the dataset compatible with PyTorch
    ``DataLoader`` and multiprocessing workers.

    Only the manifest DataFrame and lightweight metadata are stored in memory at
    initialization time. Feature arrays are loaded from disk when a sample is
    requested.

    Parameters
    ----------
    manifest_df : pandas.DataFrame
        DataFrame containing slide identifiers and labels. Must include at least
        the columns specified by ``slide_col`` and ``label_col``.

    data_dirs : list[str or pathlib.Path]
        List of directories containing the HDF5 feature files. The dataset will
        search these directories for files named <slide_id>.h5. Only the provided
        directories are searched; no nested recursion is performed.

    slide_col : str, default="slide_id"
        Name of the column in ``manifest_df`` containing slide identifiers.

    label_col : str, default="label"
        Name of the column in ``manifest_df`` containing encoded numeric labels
        (0, 1, 2, ...).

    class_col : str or None, default=None
        Optional column name in ``manifest_df`` containing human-readable class names.
        If None, class information is ignored.

    Returns
    -------
    tuple
        A tuple ``(bag, label, coords)`` where:

        - bag : torch.Tensor of shape (N, D)
            Feature bag for the given slide.

        - label : int or float
            Encoded class label corresponding to the slide.

        - coords : numpy.ndarray of shape (N, 2) or None
            Patch coordinates if available, otherwise None.
    """

    def __init__(self, 
                 manifest_df: pd.DataFrame, 
                 data_dirs: list[str | Path],
                 slide_col: str = "slide_id",
                 label_col: str = "label",
                 class_col: str | None = None):
        
        self.slide_df = manifest_df.copy().reset_index(drop=True)
        self.data_dirs = data_dirs 
        self._slide_col = slide_col
        self._label_col = label_col
        self._class_col = class_col

    def get_class_from_id(self, idx: int) -> str | None:
        if self._class_col is None:
            return None

        value = self.slide_df.iloc[idx][self._class_col]

        if pd.isna(value):
            return None

        return str(value)

    def get_class_from_slide_id(self, slide_id: str) -> str | None:
        if self._class_col is None:
            return None

        row = self.slide_df[self.slide_df[self._slide_col] == slide_id]

        if row.empty:
            raise KeyError(f"slide_id '{slide_id}' not found in dataframe")

        value = row.iloc[0][self._class_col]

        if pd.isna(value):
            return None

        return str(value)

    def _get_bag_path(self,slide_id : str):
        return find_wsi_features(slide_id,self.data_dirs)
    
    def _get_bag_vectors(self,
                         slide_id : str, 
                         features_keyword = "features", 
                         coords_keyword = "coords"):
        
        bag_h5 = self._get_bag_path(slide_id)
        with h5py.File(bag_h5, "r") as hdf5_file:
            
            _bag_h5 = hdf5_file.get(features_keyword, None)
            _coords_h5 = hdf5_file.get(coords_keyword, None)

            bag = _bag_h5[:] if _bag_h5 is not None else None
            coords = _coords_h5[:] if _coords_h5 is not None else None

        return bag, coords

    def __getitem__(self, idx):
        slide_id = self.slide_df[self._slide_col][idx]
        label = self.slide_df[self._label_col][idx]

        bag, coords = self._get_bag_vectors(slide_id= slide_id)

        bag = torch.tensor(bag)

        return bag, label, coords, slide_id# , torch.ones(bag.shape[0]) # HACK returns a mask. Uncomment when collate_fn is None. 

    def __len__(self):
        return len(self.slide_df)
```

#### Collate Function
```
# FIXME: the code inside this function accumulates/reserves/occupies several GB of memory per worker 
def mil_pad_collate(batch):
    """
    Collate function for FeatureBagsDataset that:
    - Pads all bags in a batch to the same length
    - Returns a mask indicating real vs padded entries

    Parameters
    ----------
    batch : list of tuples
        Each element is (bag, label, coords, slide_id)

    Returns
    -------
    padded_bags : torch.Tensor
        Tensor of shape (B, max_len, D)
    labels : torch.Tensor
        Tensor of shape (B,)
    masks : torch.Tensor
        Tensor of shape (B, max_len) with 1 for real entries, 0 for padding
    coords_list : list
        List of coords arrays (unaltered)
    slide_ids : list
        List of slide_ids
    """
    
    bags, labels, coords, slide_ids = zip(*batch)

    # Determine max bag length in this batch
    max_len = max(bag.shape[0] for bag in bags)

    padded_bags = []
    masks = []

    for bag in bags:
        length = bag.shape[0]
        # pad at the end along the first dimension (number of instances)
        pad_len = max_len - length
        if pad_len > 0:
            bag_padded = F.pad(bag, (0,0,0,pad_len))  # pad only first dimension
        else:
            bag_padded = bag
        mask = torch.cat([torch.ones(length, dtype=torch.float32),
                          torch.zeros(pad_len, dtype=torch.float32)])
        padded_bags.append(bag_padded)
        masks.append(mask)

    padded_bags = torch.stack(padded_bags)      # shape: (B, max_len, D)
    masks = torch.stack(masks)                  # shape: (B, max_len)
    labels = torch.tensor(labels)

    return padded_bags, labels, coords, list(slide_ids), masks
```

#### Memory Snapshot:
```
def worker_memory_report(top_k=3):
    import os
    import psutil
    import gc
    import torch
    import numpy as np
    from collections import defaultdict
    from torch.utils.data import get_worker_info


    info = get_worker_info()
    pid = os.getpid()
    ppid = os.getppid()
    proc = psutil.Process(pid)
    if info is None:
        print(f"[MAIN PROCESS] PID={pid}")
    else:
        print(f"[WORKER {info.id}] PID={pid} | num_workers={info.num_workers}")
    mem = proc.memory_info()

    rss = mem.rss / 1e6       # actual RAM used
    vms = mem.vms / 1e6       # virtual memory reserved
    shared = mem.shared / 1e6 if hasattr(mem, "shared") else 0

    # --- Scan Python objects ---
    objs = gc.get_objects()
    total_objs = len(objs)

    # Categorizzazione memoria
    category_stats = defaultdict(lambda: {"count": 0, "mem": 0})

    for o in objs:
        try:
            if isinstance(o, torch.Tensor):
                cat = "torch.Tensor"
                size = o.numel() * o.element_size()
            elif isinstance(o, torch.storage.UntypedStorage):
                cat = "torch.Storage"
                size = o.size() * o.element_size()
            elif isinstance(o, np.ndarray):
                cat = "numpy.ndarray"
                size = o.nbytes
            else:
                cat = type(o).__name__
                try:
                    size = o.__sizeof__()
                except Exception:
                    size = 0
        except Exception:
            cat = type(o).__name__
            size = 0

        category_stats[cat]["count"] += 1
        category_stats[cat]["mem"] += size

    # --- Compute totals ---
    explained_mem = sum(v["mem"] for v in category_stats.values()) / 1e6
    unexplained_mem = rss - explained_mem

    # --- Top K categories by memory ---
    top_categories = sorted(category_stats.items(), key=lambda x: x[1]["mem"], reverse=True)[:top_k]

    # --- Print report ---
    print("=" * 80)
    print(f"PROCESS MEMORY REPORT")
    print(f"PID: {pid} | PPID: {ppid}")
    print("=" * 80)

    print(f"OS MEMORY (Task Manager view)")
    print(f"RSS (used RAM)        : {rss:.2f} MB")
    print(f"VMS (reserved)        : {vms:.2f} MB")
    print(f"Shared memory         : {shared:.2f} MB")
    print("-" * 80)

    print(f"PYTHON OBJECTS TOTAL  : {total_objs}")
    print("-" * 80)

    print(f"TOP {top_k} OBJECT CATEGORIES BY MEMORY:")
    for cat, stats in top_categories:
        print(f"{cat:25} | count: {stats['count']:7} | memory: {stats['mem']/1e6:10.2f} MB")

    print("-" * 80)
    print(f"Explained memory      : {explained_mem:.2f} MB")
    print(f"Unexplained memory    : {unexplained_mem:.2f} MB")
    print("=" * 80)

def where_am_i():
    from torch.utils.data import get_worker_info
    import os
    info = get_worker_info()
    if info is None:
        print(f"[MAIN PROCESS] PID={os.getpid()}")
    else:
        print(f"[WORKER {info.id}] PID={os.getpid()} | num_workers={info.num_workers}")
```

### Extra
- I omitted the following warning for clarity (raised by  `worker_memory_report()`):
```
FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead
` rised by worker_memory_report`
```

- I tried to modify mil_pad_collate with a simplified implementation, where neither padding nor stacking operation was performed. No improvement was observed.

- The same problem is observed outside the debugging mode.

- Worker's memory snapshots were acquired through VSCode Debug Console, after fetching the 385-th datapoint in the main process (conditional breakpoint). 

### Versions

PyTorch version: 2.9.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Enterprise (10.0.26100 64 bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.19 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 16:41:31) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA RTX 4500 Ada Generation
GPU 1: NVIDIA RTX 4500 Ada Generation

Nvidia driver version: 572.16
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: Intel(R) Xeon(R) w5-2565X
Manufacturer: GenuineIntel
Family: 179
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3192
MaxClockSpeed: 3192
L2CacheSize: 36864
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] lovely-numpy==0.2.18
[pip3] mkl_fft==2.1.1
[pip3] mkl_random==1.3.0
[pip3] mkl-service==2.5.2
[pip3] numpy==1.26.4
[pip3] segmentation_models_pytorch==0.5.0
[pip3] torch==2.9.1+cu128
[pip3] torch-geometric==2.7.0
[pip3] torchvision==0.24.1+cu128
[conda] blas                           1.0              mkl
[conda] intel-openmp                   2025.0.0         haa95532_1164
[conda] lovely-numpy                   0.2.18           pypi_0                 pypi
[conda] mkl                            2025.0.0         h5da7b33_930
[conda] mkl-service                    2.5.2            py310h0b37514_0
[conda] mkl_fft                        2.1.1            py310h300f80d_0
[conda] mkl_random                     1.3.0            py310ha5e6156_0
[conda] numpy                          1.26.4           pypi_0                 pypi
[conda] numpy-base                     2.2.5            py310h7794460_2
[conda] segmentation-models-pytorch    0.5.0            pypi_0                 pypi
[conda] tbb                            2022.3.0         h90c84d6_0
[conda] tbb-devel                      2022.3.0         h90c84d6_0
[conda] torch                          2.9.1+cu128      pypi_0                 pypi
[conda] torch-geometric                2.7.0            pypi_0                 pypi
[conda] torchvision                    0.24.1+cu128     pypi_0                 pypi

cc @jerryzh168 @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov @scotts
